### ISSUE TITLE
Editor serialization

### ISSUE THREAD

**Issue by kirovboris** [Microsoft/monaco-editor#239]
monaco-editor npm version: 0.6.1
Browser: Electron
OS: Windows 10

Hi, is there any way to serialize the editing session? I want to have the capability to save the code, selection range, cursor, undo manager etc. on the server side.
I've found ```saveViewState``` method, but i can't find the way to serialize the editor model.

**Comment by alexandrudima** [comment]
At this time there's no API to serialize the undo/redo stack, so that would be lost, the view state contains the cursor state (selection state) and scroll state, you can save the model text via model.getValue(). We plan to add serializable undo/redo stack soon, but it is problematic w.r.t. what should be its limit (number of edits vs size of the edits) -- today the undo stack is unbounded which is okaysh given it sits in memory, when serializing it, we need to introduce some cut-off heuristics.

**Comment by fabiospampinato** [comment]
> when serializing it, we need to introduce some cut-off heuristics.

Maybe the burden of this could be passed on the user by requiring an argument for the maximum number of steps to preserve or something?

**Comment by michael-hawker** [comment]
Agree with @fabiospampinato even if I could get the last 100-1000 edits, it'd be better than nothing. Just have that be a configurable option and let the dev/user make the choices.

This is super painful for using this as a component of a complex application.

**Comment by michael-hawker** [comment]
Also forgot to call out that this is super important in regards to just being able to re-hydrate state of the editor if the user's work spans across application sessions. It's super handy/amazing when a developer discovers they can still undo work they did the previous time the app ran.

**Comment by chingucoding** [comment]
Any updates on this? I think being able to restore an existing session would definitely improve UX of apps by a lot since being able to pick up where you left off is super handy as @michael-hawker pointed out.

**Comment by LoganDark** [comment]
> It's super handy/amazing when a developer discovers they can still undo work they did the previous time the app ran.

Absolutely. Going to have to implement my own hacky solution for this since it's been sitting for almost 5 years. Absolutely disappointed that there is still no way to serialize the undo stack.

**Comment by 0xdevalias** [comment]
In case people on this thread haven't seen this other thread, here was a solution posted for `monaco-editor@0.36.1` in May 2023; though a followup post from February 2024 seems to imply it no longer works as is:

> This code works. Obviously, it's all private APIs, so YMMV and is only valid as of `monaco-editor@0.36.1`, although it may work in other versions too.
> 
> ```typescript
> import { editor } from 'monaco-editor';
> 
> const monacoEditor: editor.IStandaloneCodeEditor = /* Your Monaco editor instance */;
> 
> const model = monacoEditor.getModel();
> 
> const unsafeModel = model as unknown as {
>   _commandManager: {
>     _undoRedoService: {
>       createSnapshot: (resource: {}) => {};
>       restoreSnapshot: (resource: {}) => {};
>     };
>   };
> };
> 
> // Create a snapshot of the current model state
> const snapshot = unsafeModel._commandManager._undoRedoService.createSnapshot(unsafeModel._associatedResource);
> 
> // maybe serialize/store the snapshot somewhere
> 
> // Restore a previously saved snapshot
> unsafeModel._commandManager._undoRedoService.restoreSnapshot(snapshot);
> ``` 
> 
> _Originally posted by @lukeautry in [#686](https://github.com/microsoft/monaco-editor/issues/686#issuecomment-1551285660)_
